src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php

```diff
@@ -79,14 +82,38 @@ public function decode(array $encodedEnvelope): Envelope
         }
 
         try {
-            $message = $this->serializer->deserialize($encodedEnvelope['body'], $encodedEnvelope['headers']['type'], $this->format, $context);
+            $message = $this->deserialize($encodedEnvelope, $context);
         } catch (ExceptionInterface $e) {
             throw new MessageDecodingFailedException('Could not decode message: '.$e->getMessage(), $e->getCode(), $e);
         }
 
         return new Envelope($message, $stamps);
     }
 
+    private function deserialize(array $encodedEnvelope, array $context)
+    {
+        $resolver = $context[self::TYPE_RESOLVER] ?? new HeaderTypeResolver();
+
+        if ($resolver instanceof TypeResolverInterface) {
+            $type = $resolver->resolve($encodedEnvelope);
+
+            return $this->serializer->deserialize($encodedEnvelope['body'], $type, $this->format, $context);
+        }
+
+        if ($resolver instanceof DecodedAwareTypeResolverInterface) {
```

Comment:
Maybe I misunderstand what you would like, but I think it's not possible since `TypeResolverInterface` and `DecodedAwareTypeResolverInterface` have no common ancestor.

Reply:
Rewrite this line as `if (!$resolver instanceof DecodedAwareTypeResolverInterface) {` and change the body of the if-statement

Reply:
Ah ok :) Updated